### PR TITLE
Feat/sgid experimental integration

### PR DESCRIPTION
**Pull request by alexanderleegs**
This PR implements sgID login for OGP officers on a specific endpoint (`/ogp-login`). To be reviewed in conjunction with PR[#860](https://github.com/isomerpages/isomercms-backend/pull/860) on the isomercms-backend repo.

In this PR, the `LoginPage` component has been duplicated into a new `LoginPageSgid` component, which adds an additional tab for sgID login. Upon authenticating with sgID, the user should have their session details stored, and thus be redirected to the `/sites` page, similar to logging in via other means. In the case of an error occurring during the login, or the user having insufficient permissions to use this method of login (i.e. non-OGP officers), they will be redirected to `/ogp-login?status=401` or `/ogp-login?status=500`, which displays an additional error toast.

Flow diagram:
Untitled (2)

## Tests

- [ ] Login using sgid as OGP officer
- [ ] Login using other login methods from `/ogp-login`

### CODE REVIEW

File: src/layouts/Login/LoginPageSgid.tsx [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+              Privacy policy
+            </Link>
+          </Text>
+        </TabPanels>
+      </Tabs>
+    </VStack>
+  )
+}
+
+export const LoginPageSgid = (): JSX.Element => (
```
Comment (seaerchin) [inline src/layouts/Login/LoginPageSgid.tsx:215]: this is abit confusing to me - it seems like this portion is repeated code between this/`LoginPage`? why wouldn't we just use the same layout for both?

File: src/layouts/Login/LoginPageSgid.tsx [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    _hover={{
+      color: "text.link.dark",
+      textDecorationLine: "underline",
+    }}
+  >
+    {children}
+  </Link>
+)
+
+const LoginContent = (): JSX.Element => {
```
Comment (seaerchin) [inline src/layouts/Login/LoginPageSgid.tsx:76]: it seems like the crux of sgid is just the api calls, which we should just abstract out as a hook? 

the component (`LoginPageSgid`) should be removed isn't it?

File: src/routing/RouteSelector.jsx [hunk 3]
```diff
@@ -47,6 +48,12 @@ export const RouteSelector = () => (
     <Switch>
       <RedirectIfLoggedInRoute exact path="/" unauthedComponent={LoginPage} />
 
+      <RedirectIfLoggedInRoute
+        exact
+        path="/ogp-login"
```
Comment (seaerchin) [inline src/routing/RouteSelector.jsx:53]: i'm not entirely sure that this is what was intended - do we have a rfc/spec to refer to? i think this endpoint might be exposed on the backend instead?

File: src/types/login.ts [hunk 4]
```diff
@@ -5,3 +5,7 @@ export interface LoginParams {
 export interface VerifyOtpParams extends LoginParams {
   otp: string
 }
+
+export interface GetSgidAuthUrlResponseDto {
```
Comment (seaerchin) [inline src/types/login.ts:9]: can we avoid putting http methods in the type?

**Review by seaerchin — COMMENTED** [review]
this PR is abit confusing, especially to devs who might not be familiar with OAuth/the token exchange. 

perhaps what might help here is drawing a sequence diagram (suggest using [this](https://sequencediagram.org/)) and having a list of guarantees/invariants that should always hold true (<- this might be more on sgid side, eg that the same `code` cannot be reuse)

File: src/hooks/loginHooks/useGetSgidAuth.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import { useMutation, UseMutationResult } from "react-query"
+
+import * as LoginService from "services/LoginService"
+
+import { useErrorToast } from "utils/toasts"
+
+import { GetSgidAuthUrlResponseDto } from "types/login"
+import { DEFAULT_RETRY_MSG } from "utils"
+
+export const useGetSgidAuth = (): UseMutationResult<
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/hooks/loginHooks/useGetSgidAuth.ts:11]: I was wondering whether to use a GET or POST request for this actually - the call does actually modify state in our backend, namely that a unique entry is generated with details of the user's auth session - happy to discuss this more offline!

File: src/hooks/loginHooks/useGetSgidAuth.ts [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+  GetSgidAuthUrlResponseDto,
+  AxiosError,
+  void
+> => {
+  const errorToast = useErrorToast()
+  return useMutation<GetSgidAuthUrlResponseDto, AxiosError, void>(
+    () => LoginService.getSgidAuthUrl(),
+    {
+      onSuccess: (data) => {
+        window.location.assign(data.redirectUrl)
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/hooks/loginHooks/useGetSgidAuth.ts:21]: `useRedirectHook` doesn't seem to work for this! When I used it, I was getting redirected to `/` - couldn't figure out why

File: src/hooks/loginHooks/useGetSgidAuth.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      },
+      onError: () => {
+        errorToast({
+          id: "sgid-oauth-error",
+          description: `Something went wrong. ${DEFAULT_RETRY_MSG}`,
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/hooks/loginHooks/useGetSgidAuth.ts]: You're right, changed in https://github.com/isomerpages/isomercms-frontend/pull/1369/commits/26177c7f4523132e8a751e141f56d66b79a3a030

File: src/layouts/Login/LoginPageSgid.tsx [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+  const { search } = useLocation()
+
+  const [statusCode, setStatusCode] = useState("200")
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/layouts/Login/LoginPageSgid.tsx]: Removed in https://github.com/isomerpages/isomercms-frontend/pull/1369/commits/857b7eadbc821cea2f097ff589df87b6daa148a4!

File: src/layouts/Login/LoginPageSgid.tsx [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+  useEffect(() => {
+    const params = new URLSearchParams(search)
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/layouts/Login/LoginPageSgid.tsx]: `useParams` is for dynamic params in the route, but we're looking for query params here!

File: src/layouts/Login/LoginPageSgid.tsx [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+    {children}
+  </Link>
+)
+
+const LoginContent = (): JSX.Element => {
+  const { search } = useLocation()
+
+  const [statusCode, setStatusCode] = useState("200")
+
+  useEffect(() => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/layouts/Login/LoginPageSgid.tsx]: Yeah that's a good point, changed in https://github.com/isomerpages/isomercms-frontend/pull/1369/commits/857b7eadbc821cea2f097ff589df87b6daa148a4

File: src/layouts/Login/LoginPageSgid.tsx [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+        break
+      default:
+        errorMessage = `Something went wrong. ${DEFAULT_RETRY_MSG}`
+    }
+    errorToast({
+      id: "sgid-login-error",
+      title: "Error",
+      description: errorMessage,
+    })
+  }, [errorToast, statusCode])
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply src/layouts/Login/LoginPageSgid.tsx]: Yup, so we wouldn't be able to get error messages as per normal

File: src/layouts/Login/LoginPageSgid.tsx (hunk 2 above)
Comment (alexanderleegs) [reply src/layouts/Login/LoginPageSgid.tsx]: https://github.com/isomerpages/isomercms-frontend/pull/1369/commits/b922ca67e5bd7957924ad4a60ee74dff4dbadf1a

File: src/routing/RouteSelector.jsx (hunk 3 above)
Comment (alexanderleegs) [reply src/routing/RouteSelector.jsx]: This is intended, we have a separate login endpoint (`/ogp-login`) which ogp officers can use for now, so as not to confuse users about the availability of this functionality

File: src/types/login.ts (hunk 4 above)
Comment (alexanderleegs) [reply src/types/login.ts]: https://github.com/isomerpages/isomercms-frontend/pull/1369/commits/a14be908614d8a13dd3107dfbff4f5e276affcf1

**Review by QiluXie — APPROVED** [review]
Approved based on manual testing in staging for experimental release

**Pull request merged**